x/coostaking/keeper/intercept_fee_collector.go

```diff
@@ -27,13 +32,72 @@ func (k Keeper) HandleCoinsInFeeCollector(ctx context.Context) error {
 		return nil
 	}
 
-	coostakingPortion := k.GetParams(ctx).CoostakingPortion
-	coostakingRewards := ictvtypes.GetCoinsPortion(feesCollectedInt, coostakingPortion)
-	return k.AccumulateCoostakingRewards(ctx, coostakingRewards)
+	params := k.GetParams(ctx)
+
+	valDirectRwds := ictvtypes.GetCoinsPortion(feesCollectedInt, params.ValidatorsPortion)
+	if err := k.allocateValidatorsRewards(ctx, valDirectRwds); err != nil {
+		return err
+	}
+
+	coostakingRewards := ictvtypes.GetCoinsPortion(feesCollectedInt, params.CoostakingPortion)
+	return k.accumulateCoostakingRewards(ctx, coostakingRewards)
+}
+
+// allocateValidatorsRewards allocates rewards to validators based on their voting power
+// It transfers the rewards to the distribution module account and updates the validator's commission.
+func (k Keeper) allocateValidatorsRewards(ctx context.Context, rwds sdk.Coins) error {
+	if !rwds.IsAllPositive() {
+		return nil
+	}
+	// parse rewards to decCoins
+	valsRwds := sdk.NewDecCoinsFromCoins(rwds...)
+	goCtx := sdk.UnwrapSDKContext(ctx)
+	bondedVotes := goCtx.VoteInfos()
+	// determine the total power signing the block
+	var totalPwr int64
+	for _, voteInfo := range bondedVotes {
+		totalPwr += voteInfo.Validator.Power
+	}
+
+	// safety check
+	if totalPwr == 0 {
+		return nil
+	}
+
+	// Transfer rewards to the distribution module account
+	// 'cause these direct rewards will be allocated to the validators commission
+	err := k.bankK.SendCoinsFromModuleToModule(ctx, k.feeCollectorName, distrtypes.ModuleName, rwds)
+	if err != nil {
+		return err
+	}
+
+	for _, vote := range bondedVotes {
+		validator, err := k.stkK.ValidatorByConsAddr(ctx, vote.Validator.Address)
+		if err != nil {
+			return err
+		}
+
+		powerFraction := math.LegacyNewDec(vote.Validator.Power).QuoTruncate(math.LegacyNewDec(totalPwr))
```

Comment:
```suggestion
    totalPwrDec := math.LegacyNewDec(totalPwr)
	for _, vote := range bondedVotes {
		validator, err := k.stkK.ValidatorByConsAddr(ctx, vote.Validator.Address)
		if err != nil {
			return err
		}

		powerFraction := math.LegacyNewDec(vote.Validator.Power).QuoTruncate(totalPwrDec)
```

suggestion to move total power outside for loop